src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java

```diff
@@ -32,76 +33,88 @@
 public class InMemoryCommandLog implements CommandLog {
 
     private final DirectFactory directFactory = new AllocatingDirectFactory();
+
     private final List<LogEntry> entries = new ArrayList<>();
+    private final Long2LongHashMap maxCommandIndexBySourceId = new Long2LongHashMap(-1);
 
     @Override
-    public long size() {
+    public synchronized long size() {
         return entries.size();
     }
 
     @Override
-    public synchronized int readTerm(final long position) {
-        return read(position, null).logKey().term();
-    }
-
-    @Override
-    public synchronized LogEntry read(final long position, final DirectFactory directFactory) {
-        return clone(read(position));
+    public synchronized int readTerm(final long index) {
+        return read(index).logKey().term();
     }
 
     @Override
-    public void readTo(final long position, final LogEntry logEntry) {
-        logEntry.copyFrom(read(position));
+    public synchronized void readKeyTo(final long index, final LogKey logKey) {
```

Comment:
in the interface: LogKey target
shouldn't they be consistent? or targetLogKey for clarity